File: problems/851_loud-and-rich/main.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution
{
public:
    int dfs(int u, vector<vector<int>> &graph, vector<int> &quiet, vector<int> &ans)
    {
        if (ans[u] != -1)
            return ans[u];
        ans[u] = u;
        for (int v : graph[u])
        {
            int t = dfs(v, graph, quiet, ans);
            if (quiet[t] < quiet[ans[u]])
                ans[u] = t;
        }
        return ans[u];
    }

    vector<int> loudAndRich(vector<vector<int>> &richer, vector<int> &quiet)
    {
        int n = quiet.size();
        vector<int> ans(n, -1);
        vector<vector<int>> graph(n);

        for (auto &e : richer)
            graph[e[1]].push_back(e[0]);

        for (int i = 0; i < n; i++)
            if (ans[i] == -1)
                dfs(i, graph, quiet, ans);

        return ans;
    }

    vector<int> loudAndRichTopological(vector<vector<int>> &richer, vector<int> &quiet)
    {
        int n = quiet.size();
        vector<int> res(n);
        if (n == 0)
            return res;
        vector<vector<int>> graph(n);
        vector<int> indeg(n, 0);
        for (int i = 0; i < richer.size(); i++)
        {
            graph[richer[i][0]].push_back(richer[i][1]);
            indeg[richer[i][1]]++;
        }

        for (int i = 0; i < n; ++i)
            res[i] = i;
        queue<int> q;
        for (int i = 0; i < n; i++)
            if (indeg[i] == 0)
                q.push(i);
        while (!q.empty())
        {
            int curr = q.front();
            q.pop();
            for (auto nei : graph[curr])
            {
                if (quiet[res[curr]] < quiet[res[nei]])
                {
                    res[nei] = res[curr];
                }
                if (--indeg[nei] == 0)
                    q.push(nei);
            }
        }

        return res;
    }
};

void printVector(const vector<int> &values)
{
    cout << "[";
    for (size_t i = 0; i < values.size(); ++i)
    {
        cout << values[i];
        if (i + 1 < values.size())
            cout << ", ";
    }
    cout << "]";
}

int main()
{
    Solution sol;

    vector<vector<int>> richer1 = {{1, 0}, {2, 1}, {3, 1}, {3, 7}, {4, 3}, {5, 3}, {6, 3}};
    vector<int> quiet1 = {3, 2, 5, 4, 6, 1, 7, 0};
    vector<vector<int>> richer2 = {};
    vector<int> quiet2 = {0};

    auto result1 = sol.loudAndRich(richer1, quiet1);
    cout << "Sample 1 placeholder result: ";
    printVector(result1);
    cout << "\nSample 1 expected output: [5,5,2,5,4,5,6,7]\n\n";

    auto result2 = sol.loudAndRich(richer2, quiet2);
    cout << "Sample 2 placeholder result: ";
    printVector(result2);
    cout << "\nSample 2 expected output: [0]\n";

    return 0;
}

File: problems/851_loud-and-rich/main.rs
struct Solution;

impl Solution {
    pub fn loud_and_rich(richer: Vec<Vec<i32>>, quiet: Vec<i32>) -> Vec<i32> {
        let n = quiet.len();
        let mut graph: Vec<Vec<usize>> = vec![vec![]; n];
        for r in richer {
            let a = r[0] as usize;
            let b = r[1] as usize;
            graph[b].push(a); // b -> a, a is richer than b
        }
        let mut ans = vec![-1i32; n];
        for i in 0..n {
            if ans[i] == -1 {
                Self::dfs(i, &graph, &quiet, &mut ans);
            }
        }
        ans
    }

    fn dfs(u: usize, graph: &Vec<Vec<usize>>, quiet: &Vec<i32>, ans: &mut Vec<i32>) -> i32 {
        if ans[u] != -1 {
            return ans[u];
        }
        ans[u] = u as i32;
        for &v in &graph[u] {
            let t = Self::dfs(v, graph, quiet, ans);
            if quiet[t as usize] < quiet[ans[u] as usize] {
                ans[u] = t;
            }
        }
        ans[u]
    }

    pub fn loud_and_rich_topological(richer: Vec<Vec<i32>>, quiet: Vec<i32>) -> Vec<i32> {
        let n = quiet.len();
        let mut graph: Vec<Vec<usize>> = vec![vec![]; n];
        let mut indeg = vec![0; n];
        for r in &richer {
            let a = r[0] as usize;
            let b = r[1] as usize;
            graph[a].push(b);
            indeg[b] += 1;
        }
        let mut res: Vec<i32> = (0..n as i32).collect();
        use std::collections::VecDeque;
        let mut q = VecDeque::new();
        for i in 0..n {
            if indeg[i] == 0 {
                q.push_back(i);
            }
        }
        while let Some(curr) = q.pop_front() {
            for &nei in &graph[curr] {
                if quiet[res[curr] as usize] < quiet[res[nei] as usize] {
                    res[nei] = res[curr];
                }
                indeg[nei] -= 1;
                if indeg[nei] == 0 {
                    q.push_back(nei);
                }
            }
        }
        res
    }
}

fn main() {
    let richer1 = vec![
        vec![1, 0],
        vec![2, 1],
        vec![3, 1],
        vec![3, 7],
        vec![4, 3],
        vec![5, 3],
        vec![6, 3],
    ];
    let quiet1 = vec![3, 2, 5, 4, 6, 1, 7, 0];
    let richer2: Vec<Vec<i32>> = vec![];
    let quiet2 = vec![0];

    let result1 = Solution::loud_and_rich(richer1, quiet1);
    println!("Sample 1 placeholder result: {:?}", result1);
    println!("Sample 1 expected output: [5,5,2,5,4,5,6,7]\n");

    let result2 = Solution::loud_and_rich(richer2, quiet2);
    println!("Sample 2 placeholder result: {:?}", result2);
    println!("Sample 2 expected output: [0]");
}
